sound/soc/sof/intel/hda-dsp.c

```diff
@@ -799,11 +799,15 @@ int hda_dsp_runtime_idle(struct snd_sof_dev *sdev)
 
 int hda_dsp_runtime_suspend(struct snd_sof_dev *sdev)
 {
+	struct sof_intel_hda_dev *hda = sdev->pdata->hw_pdata;
 	const struct sof_dsp_power_state target_state = {
 		.state = SOF_DSP_PM_D3,
 	};
 	int ret;
 
+	/* cancel any attempt for DSP D0I3 */
+	cancel_delayed_work_sync(&hda->d0i3_work);
```

Comment:
what about system suspend, isn't it required as well? Otherwise we'd enter D0i3 upon resume...

Reply:
we already do it during system suspend. It seems to have been missed for runtime suspend.

Reply:
you probably need a Fixes tag, that's pretty major and would need to be back-ported. If that's handled with the -stable kernel it's easier for everyone.

We had similar problems with SoundWire, the rule is that all delayed work shall be cancelled... not doing it will invariably lead to disasters.

Reply:
done @plbossart